Upgrade to pageres@3

SamVerschueren opened the pull request:
Submitting my changes so far because I'm a little bit stuck here and it would be easier to discuss if we had something we can work with :).

I added meow because the test that tests if the help data is printed was not working. With meow it did, but that wasn't the reason why it isn't working.

The reason why all tests are stuck is because the program doesn't get past the `getStdin()` method at line [209](https://github.com/sindresorhus/pageres-cli/blob/iss216/cli.js#L209). The promise is never resolved or rejected, so it doesn't continue to generate the pages.

Any ideas?

// @sindresorhus @kevva

sindresorhus commented:
I think I know the problem. All the tests have to use `stdio: [process.stdin, null, null]`, otherwise it will try to read from stdin as `process.stdin.isTTY` will be false.

SamVerschueren commented:
But the only active test (generate screenshot) uses `stdio: [process.stdin, null, null]` and it still fails to work. The others are commented out.

sindresorhus commented:
Ok, I know why. In AVA 0.3.0 we now spawn the test files, so that won't work, since `process.stdin.isTTY` is undefined and is then coerced to false and activates the stdin read. I guess we could either remove the stdin ability (does anyone actually use that anyway?) or add a stdin flag. Preferably former.

SamVerschueren commented:
I removed `stdin` and fixed all the tests. We might still offer it (without the ability of testing it) if we write this.

```
if (process.stdin.isTTY === false) {
	getStdin().then(function (data) {
		[].push.apply(args, data.trim().split(/\r?\n/));
		init(args, options);
	});
} else {
	init(args, options);
}
```

Not sure though, not very familiar with that.

SamVerschueren commented on readme.md:
```diff
@@ -87,9 +87,6 @@ pageres --delay 3 1366x768 < urls.txt
 pageres unicorn.html 1366x768
-
-# Pipe in resolutions
-cat screen-resolutions.txt | pageres todomvc.com yeoman.io
```
I guess this isn't possible anymore by removing the `stdin` stuff?

sindresorhus replied:
Yes, but I always found it awkward tbh. Better to just use string interpolation in the command.

SamVerschueren commented:
I'm ok with removing it :). Just wanted to point out a solution if you still wanted it, which is not the case so let's just drop it. If people want it, they can always install an older version...

SamVerschueren commented:
Weird, your comment keeps saying `commented just now`

sindresorhus commented:
Yes, there are many ways to still support it. We could even have just used an environment variable for testing and checked it in the CLI. But I don't see the value of the stdin feature and want to just remove it.

sindresorhus commented:
Yeah, I noticed too. Super weird.

sindresorhus commented:
And I commented before you.

SamVerschueren commented:
I was thinking of creating a `get-child-process` (or `get-spawn`) module that does the same as [get-stream](https://github.com/sindresorhus/get-stream) but for a spawned child processes. Tests would become more clean but not sure if it's usefull in other cases as well. Wdyt?

sindresorhus commented:
Why not just `get-stream` directly? `stdout` is just a stream.

SamVerschueren commented:
Used `get-stream` now in some tests.

But if we create a new module we could maybe do it like this.

```js
test.serial('generate screenshot', t => {
	await getSpawn(spawn('../cli.js', ['yeoman.io', '320x240']));

	t.true(pathExists.sync('yeoman.io-320x240.png'));
	fs.unlinkSync('yeoman.io-320x240.png');
});
```

The promise resolves the output of `cp.stdout` and rejects the output of `cp.stderr`. It abstracts away the underlying streams which results in cleaner tests I guess :).

```js
test('show error if no url is specified', async t => {
	t.throws(getSpawn(spawn('../cli.js', ['320x240'])), Error);
});
```

Just some ideas :).

sindresorhus commented:
Good point. Could even go further:

```js
await getSpawn('../cli.js', ['yeoman.io', '320x240']));
```

But you should use `execFile` underneath, which makes the name weird, maybe `get-cp` instead?

SamVerschueren commented:
Would call it `get-child-process`. Says more then `get-cp`, no?

```js
const getCp = require('get-child-process');
```

Why `execFile` underneath? Not sure which one (execFile or spawn) to use when...

SamVerschueren commented:
Or wrap both?

```js
const cp = require('get-child-process');

cp.execFile('../cli.js').then(data => {});

cp.spawn('../cli.js').then(...);
```

sindresorhus commented:
`execFile` is the buffered version of `spawn`.

> Or wrap both?

:+1: 

-

Make sure you return both `stdout` and `stderr` in an object.

We can then use it like:

```js
const {stdout} = await cp.execFile('../cli.js', ['yeoman.io', '320x240']);
```

sindresorhus commented:
Actually, not sure why you would need `spawn` though, as you want the buffered content, which is `execFile`.

SamVerschueren commented:
Why return an object with stdout and stderr? Just reject the promise if an error occurred?

`get-exec-file` then :p?

sindresorhus commented:
> Why return an object with stdout and stderr? Just reject the promise if an error occurred?

Because `stderr` can sometimes be useful without an error.

> get-exec-file then :p?

Yeah, sure. Doesn't really matter.

SamVerschueren commented:
Oh oké, I was thinking that if `stderr` emits data, reject allways because an error occurred. But that might be too naieve then?

sindresorhus commented:
> Oh oké, I was thinking that if stderr emits data, reject allways because an error occurred. But that might be too naieve then?

Yes, as I said earlier, stderr isn't necessarily connected to an error, it's just an alternative stream commonly used for displaying errors, but it can be used for anything. Funnily enough, we actually do that here in pageres. Since stdout is used for transferring the screenshot, we use stderr for everything else like logging: https://github.com/kevva/screenshot-stream/blob/ef4a4de2eedfc6c86d918ffa46873f2dc8ecfe6f/index.js#L63

SamVerschueren commented:
I guess this is it: [get-exec-file](https://github.com/SamVerschueren/get-exec-file). Feel free to provide feedback on the module if you think some things are missing.

Added it to the tests of `pageres-cli` which are much cleaner now. The only test I still had to use `spawn` is the `remove temporary files on cancel`.

SamVerschueren commented:
Anything I should do before it can get merged in? Or too busy with other stuff :)?

sindresorhus commented on cli.js:
```diff
@@ -3,8 +3,6 @@
 var multiline = require('multiline');
 var updateNotifier = require('update-notifier');
-var getStdin = require('get-stdin');
-var subarg = require('subarg');
```
We need that one for parsing subcommands, like `$ pageres [ foo.com 1024x100 ] [bar.com 800x600]`

sindresorhus commented on cli.js:
```diff
@@ -184,18 +175,8 @@ function parse(args, globalOptions) {
-
 	if (args.length === 0) {
-		showHelp();
+		console.log(cli.help);
```
`cli.showHelp(1);`

sindresorhus commented on test/test.js:
```diff
@@ -1,135 +1,61 @@
+	try {
+		const {stderr} = await execFile('../cli.js', ['320x240']);
+	} catch (err) {
+		t.regexTest(/Specify a url/, err.stderr);
```
You can just use `t.throws` as it supports promises now, and `err.toString()` should contain the error.

sindresorhus commented:
`get-exec-file` looks ok.

SamVerschueren commented on test/test.js:
```diff
@@ -1,135 +1,57 @@
-		t.assert(/Specify a url/.test(data), data);
-	});
+test('show error if no url is specified', async t => {
+	t.throws(execFile('../cli.js', ['320x240']), /Specify a url/);
```
It looks like there is a bug in AVA. Whatever I pass in as second argument (in this case `/Specify a url/`), the test always succeeds. Will create a new issue there.

SamVerschueren replied:
It looks like the test isn't being executed, guess I'm not using it correctly.

I define the test as `async`, but `await` is not used which is wrong so that's why he is not waiting for the `execFile` to complete.
The first argument expects a promise, so I should not use `await execFile()` as first argument because that would already resolve the outcome.
So I tried `await t.throws()` as `throws` returns the promise but it looks like that isn't working either.

sindresorhus replied:
Probably because the async functions is not aware of the promise. I guess for this to work we would have to return the promise in the `t.throws` so we could do `await t.throws`. For now though, you can just make it a normal function and do:

```
t.plan(1);
t.throws(execFile('../cli.js', ['320x240']), /Specify a url/);
```

sindresorhus replied:
// @vdemedes Thoughts?

SamVerschueren replied:
It returns the promise

https://github.com/sindresorhus/ava/blob/master/lib/assert.js#L75

vdemedes replied:
Weird, some guy already submitted a similar bug to AVA and I pushed a PR to fix it. @SamVerschueren could you please try to create a reproducible test case with only the basics, so that I can fix it?

SamVerschueren replied:
Will do if I have little bit more time. Fixed it by using `t.plan()` for now.

Is this the correct syntax?

```
await t.throws(execFile('../cli.js', ['320x240']), /Specify a url/);
```

SamVerschueren replied:
I did some further investigation on this because the AVA code seemed correct and I tested it with a more simpler example.

```js
await t.throws(test(), /hello world/);

async function test() {
	await Promise.reject(new Error('hello world'));
}
```

If the error message is changed to something else, the test fails.

-----------
Back to the `execFile` problem.

For some bizar reason, this succeeds

```js
await t.throws(test(), /hello world/);

async function test() {
	await execFile('../cli.js', ['320x240']);
}
```

So maybe there is an issue in the [get-exec-file](https://github.com/SamVerschueren/get-exec-file/blob/master/index.js) implementation... 

If I change [these lines](https://github.com/SamVerschueren/get-exec-file/blob/master/index.js#L9-L20) to this:

```js
return new Promise(function (resolve, reject) {
	reject(new Error('my error'));
});
```

The test fails again because `hello world` is not the same as `my error`. But if I change the rejection error inside the `execFile` statement

```js
return new Promise(function (resolve, reject) {
	execFile(file, args, options, function (err, stdout, stderr) {
		if (err) {
			reject(new Error('my error'));
			return;
		}

		resolve({stdout: stdout, stderr: stderr});
	});
});
```

The test succeeds again. What am I missing here?

SamVerschueren replied:
Somehow, the promise is being rejected/resolved before `execFile` is completed.

```js
execFile(file, args, options, function (err, stdout, stderr) {
	console.log('output');

	if (err) {
		err.stdout = stdout;
		err.stderr = stderr;
		reject(err);
		return;
	}

	resolve({stdout: stdout, stderr: stderr});
});
```

The `output` log is not even printed if an error occurs.

SamVerschueren replied:
It looks like something is wrong in `execFile`. If I use `pify` to implement `get-exec-file`, same things happen.

SamVerschueren commented on cli.js:
```diff
@@ -221,11 +202,4 @@ function init(args, options) {
-		init(args, options);
-	});
-}
+init(subarg(cli.input, options)._, cli.flags);
```
Do I have to pass in the `options` object (declared at the beginning of the file) in here?

sindresorhus replied:
Unsure. I don't remember how the subarg API looks like. Maybe check its docs.

SamVerschueren replied:
The docs say you can, it's the `minimist` options object. But the options object defined by us is only defining "global" arguments, not the arguments that can be passed in per sub command. At least, that's what it looks like to me :). So I think we just might leave it out here or define a second `minimist` object that defines the arguments of the subcommand.

sindresorhus replied:
No, `crop` and `delay` can be passed into each subcommand. See: https://github.com/sindresorhus/pageres-cli/blob/eabc83255c92a7b2ec7b4769b493883512811f37/cli.js#L57

SamVerschueren replied:
Oh ok, will leave it like this then :).

SamVerschueren commented:
> `get-exec-file` looks ok.

Should've called it `execify` I guess, it's more a promise wrapper around `execFile`.

sindresorhus commented:
> promise wrapper around execFile

Don't say that. I dread promise wrappers :p

SamVerschueren commented:
Better :p?

sindresorhus commented:
Haha, yes :p

sindresorhus commented:
@SamVerschueren Now that AVA is finally fixed is there anything blocking this?

SamVerschueren commented:
@sindresorhus Sorry, will look at it wednesday. I'm off to Berlin now for 3 days (city trip). Will keep you posted!

sindresorhus commented:
No rush. Enjoy :)

SamVerschueren commented:
Thanks, definitely will :).

SamVerschueren commented:
@sindresorhus All tests succeed, so I guess this can be merged?

sindresorhus commented:
Yay! Glad we could finally land this. Super good work @SamVerschueren :)

SamVerschueren commented:
Wooohoow!
